forge/src/main/java/me/lucko/luckperms/forge/LPForgeBootstrap.java

```diff
@@ -0,0 +1,265 @@
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.players.PlayerList;
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.server.ServerAboutToStartEvent;
+import net.minecraftforge.event.server.ServerStoppingEvent;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.ModContainer;
+import net.minecraftforge.fml.ModList;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
+import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
+import net.minecraftforge.fml.loading.FMLPaths;
+import net.minecraftforge.forgespi.language.IModInfo;
+import org.apache.logging.log4j.LogManager;
+import org.apache.maven.artifact.versioning.ArtifactVersion;
+
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Bootstrap plugin for LuckPerms running on Forge.
+ */
+@Mod(value = LPForgeBootstrap.ID)
+public final class LPForgeBootstrap implements LuckPermsBootstrap {
+    public static final String ID = "luckperms";
+
+    /**
+     * The plugin logger
+     */
+    private final PluginLogger logger;
+
+    /**
+     * A scheduler adapter for the platform
+     */
+    private final SchedulerAdapter schedulerAdapter;
+
+    /**
+     * The plugin class path appender
+     */
+    private final ClassPathAppender classPathAppender;
+
+    /**
+     * The plugin instance
+     */
+    private final LPForgePlugin plugin;
+
+    /**
+     * The time when the plugin was enabled
+     */
+    private Instant startTime;
+
+    // load/enable latches
+    private final CountDownLatch loadLatch = new CountDownLatch(1);
+    private final CountDownLatch enableLatch = new CountDownLatch(1);
+
+    /**
+     * The Minecraft server instance
+     */
+    private MinecraftServer server;
+
+    public LPForgeBootstrap() {
+        this.logger = new Log4jPluginLogger(LogManager.getLogger(LPForgeBootstrap.ID));
+        this.schedulerAdapter = new ForgeSchedulerAdapter(this);
+        this.classPathAppender = new ForgeClassPathAppender(this.logger);
+        this.plugin = new LPForgePlugin(this);
+
+        FMLJavaModLoadingContext.get().getModEventBus().addListener(this::onCommonSetup);
+        MinecraftForge.EVENT_BUS.register(this);
+    }
+
+    // provide adapters
+
+    @Override
+    public PluginLogger getPluginLogger() {
+        return this.logger;
+    }
+
+    @Override
+    public SchedulerAdapter getScheduler() {
+        return this.schedulerAdapter;
+    }
+
+    @Override
+    public ClassPathAppender getClassPathAppender() {
+        return this.classPathAppender;
+    }
+
+    // lifecycle
```

Comment:
Do the lifecycle listeners support the integrated server correctly?

I think the fabric module is explicit about only enabling for the dedicated server - that's fine with me, but I think the code should enforce it to prevent any confusion.

Reply:
Yes, I've tested running LuckPerms on the integrated server and it worked fine, I can make it server side only if you'd prefer.

Reply:
Nope, this is fine :)